shrine/lib/model/app_state_model.dart

```diff
@@ -0,0 +1,107 @@
+// Copyright 2018-present the Flutter authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'package:scoped_model/scoped_model.dart';
+
+import 'product.dart';
+import 'products_repository.dart';
+
+double _salesTaxRate = 0.06;
+double _shippingCostPerItem = 7.0;
+
+class AppStateModel extends Model {
+  // All the available products.
+  List<Product> _availableProducts;
+
+  // The currently selected category of products.
+  Category _selectedCategory = Category.all;
+
+  // The IDs and quantities of products currently in the cart.
+  Map<int, int> _productsInCart = {};
+
+  Map<int, int> get productsInCart => Map.from(_productsInCart);
+
+  // Total number of items in the cart.
+  int get totalCartQuantity => _productsInCart.values.fold(0, (v, e) => v + e);
+
+  Category get selectedCategory => _selectedCategory;
+
+  // Totaled prices of the items in the cart.
+  double get subtotalCost => _productsInCart.keys
+      .map((id) => _availableProducts[id].price * _productsInCart[id])
+      .fold(0.0, (sum, e) => sum + e);
+
+  // Total shipping cost for the items in the cart.
+  double get shippingCost =>
+      _shippingCostPerItem *
+      _productsInCart.values.fold(0.0, (sum, e) => sum + e);
+
+  // Sales tax for the items in the cart
+  double get tax => subtotalCost * _salesTaxRate;
+
+  // Total cost to order everything in the cart.
+  double get totalCost => subtotalCost + shippingCost + tax;
+
+  // Returns a copy of the list of available products, filtered by category.
+  List<Product> getProducts() {
+    if (_availableProducts == null) return List<Product>();
+
+    if (_selectedCategory == Category.all) {
+      return List.from(_availableProducts);
+    } else {
+      return _availableProducts.where((p) => p.category == _selectedCategory);
+    }
+  }
+
+  // Adds a product to the cart.
+  void addProductToCart(int productId) {
+    if (!_productsInCart.containsKey(productId)) {
+      _productsInCart[productId] = 1;
+    } else {
+      _productsInCart[productId]++;
+    }
+
+    notifyListeners();
+  }
+
+  // Removes an item from the cart.
+  void removeItemFromCart(int productId) {
+    if (_productsInCart.containsKey(productId)) {
+      if (_productsInCart[productId] == 1) {
+        _productsInCart.remove(productId);
+      } else {
+        _productsInCart[productId] -= 1;
```

Comment:
Might look nice to have `--` to match the one above. Or have the one above be `+=`.

Reply:
Done.